### PR TITLE
Code deploy build

### PR DESCRIPTION
**Pull request by igoristic**

### CODE REVIEW

File: cicd-scripts/app_install.sh
Comment (selfdanielj) [inline cicd-scripts/app_install.sh:45]: why do you install virtualenv here? doesnt look like you use it?

File: cicd-scripts/app_install.sh
Comment (selfdanielj) [inline cicd-scripts/app_install.sh:47]: why do you cd here?  isn't the code in `/home/ec2-user/spider`?

File: buildspec_spider.yml
Comment (selfdanielj) [inline buildspec_spider.yml:1]: what is the purpose of this file?  it seems to do a lot of what the cicd scripts do.  Do we need both?

File: cicd-scripts/app_start.sh
Comment (selfdanielj) [inline cicd-scripts/app_start.sh:14]: This should be starting scrapyd, scrapydweb, and logparser instead of scrapy directly right?

File: cicd-scripts/app_start.sh
Comment (selfdanielj) [inline cicd-scripts/app_start.sh:19]: do you think we will want to capture these logs in cloudwatch?  if so could stdout and stderr be redirected somewhere else?

File: cicd-scripts/app_install.sh
Comment (selfdanielj) [inline cicd-scripts/app_install.sh:50]: did you install the venv in a different directory on purpose ("app" vs "spider")?  i dont think it matters just curious.

File: cicd-scripts/app_start.sh
Comment (selfdanielj) [reply cicd-scripts/app_start.sh:19]: If so, im wondering if as a follow on we need some sort of logrotate.d or similar process to handle these.

File: appspec.yml
Comment (selfdanielj) [inline appspec.yml:24]: does this work without a fully resolved path? i don't know what the cwd is but probably not the project root

**Review by selfdanielj — COMMENTED** [review]
A few questions below... if any change suggestions are valid, ill leave it up to you to decide if any should be included in this story or should be done in a follow on story.

**Comment by igoristic** [comment]
@selfdanielj Thanks for all the feedback! Looks like I had some bad auto resolve conflicts with the multiple branches I force merged into 🤦‍♂️

File: cicd-scripts/app_install.sh
Comment (lsamuels-fearless) [inline cicd-scripts/app_install.sh:1]: The version of Python is hard-coded here. is there a way for us to control this in a way that wouldn't require an app-side code change?

File: cicd-scripts/app_install.sh
Comment (selfdanielj) [inline cicd-scripts/app_install.sh:7]: What about removing the environment name from this variable so we don't have to have two environment variables, SPIDER_STAGING_URLS_API and, i assume SPIDER_PROD_URLS_API when we can just have one SPIDER_URLS_API and set the value based on environment?

File: cicd-scripts/app_install.sh
Comment (selfdanielj) [inline cicd-scripts/app_install.sh:38]: does this mean we still have to make changes here for minor version updates?  Why not include the ".0" in the variable and the just use `python3` or `python` instead of `python 3.12` when upgrading pip below.

File: cicd-scripts/app_install.sh
Comment (selfdanielj) [inline cicd-scripts/app_install.sh:58]: i don't see where we are using virtualenv.  Below you are using venv the built in module.  I don't think we need this installed.

**Review by selfdanielj — APPROVED** [review]
not sure if you were waiting on an approval but ok with me to merge whenever you are ready... you did a lot of great work here! 🎉

File: search_gov_crawler/search_gov_spiders/pipelines.py
Comment (selfdanielj) [inline search_gov_crawler/search_gov_spiders/pipelines.py:76]: I think you can probably access a logger here... can this be a `logger.info()` instead of a `print`?

File: search_gov_crawler/search_gov_spiders/pipelines.py
Comment (selfdanielj) [inline search_gov_crawler/search_gov_spiders/pipelines.py:78]: similar here, can this be `logger.error()`?

File: search_gov_crawler/search_gov_spiders/pipelines.py
Comment (selfdanielj) [inline search_gov_crawler/search_gov_spiders/pipelines.py:70]: you talked about issues with staging being down and this causing lots of errrors in the logs... did you consider not catching the error here but instead raising it and failing the crawl?  Is there benefit it continuing if we can't write to this API?

File: search_gov_crawler/search_gov_spiders/pipelines.py
Comment (selfdanielj) [inline search_gov_crawler/search_gov_spiders/pipelines.py:1]: is it possible to add any tests for this new functionality either through updating the full crawl tests or adding unit tests for the class?  Even just the private methods if you don't want to mock the post response.

File: cicd-scripts/app_install.sh
Comment (igoristic) [reply cicd-scripts/app_install.sh:38]: This is based on semantic versioning, and some places in the code require only the major and a minor, eg: `3.12` If my variable also has the patch suffix `.0` then I would have to remove it where ever it's incompatible. So, basically it's easier to add the patch version rather than removing it. I didn't give it much thought though, because usually patch only has minor fixes/improvements and no breaking changes

File: cicd-scripts/app_install.sh
Comment (igoristic) [reply cicd-scripts/app_install.sh:58]: Its aliased as `venv`. It might already be installed (usually it's part of python), but it's not guaranteed

Comment (selfdanielj) [reply cicd-scripts/app_install.sh:58]: Did you alias it somewhere or is that built in functionality? venv is included in the standard python library.  

https://docs.python.org/3.12/library/venv.html
vs
https://pypi.org/project/virtualenv/

and from https://virtualenv.pypa.io/en/latest/:

> virtualenv is a tool to create isolated Python environments. Since Python 3.3, a subset of it has been integrated into the standard library under the [venv module](https://docs.python.org/3/library/venv.html). The venv module does not offer all features of this library, to name just a few more prominent.

File: search_gov_crawler/search_gov_spiders/pipelines.py
Comment (igoristic) [reply search_gov_crawler/search_gov_spiders/pipelines.py:1]: Unit test's for bash scripts are not really conventional, since you'd have to mock most of the functionality and it's environment. Which makes it's really easy to give it a "happy path" where it passes the test, but would still be a false positive under a more realistic scenario

File: cicd-scripts/app_install.sh
Comment (igoristic) [reply cicd-scripts/app_install.sh:58]: Ah you're right. I always thought they were the same thing and that we can install python without it and then add it via virtualenv

File: cicd-scripts/app_start.sh
Comment (selfdanielj) [inline cicd-scripts/app_start.sh:52]: This means the spider app won't start if the URLs API is not available right?  Nothing in the scrapyd or scrapydweb processes themselves rely on the URLs API, its just the actually running jobs.  Should we just fail the job if the API is down? Why do we have to not even start these services?

**Comment by selfdanielj** [comment]
Not sure if you've worked more on this but to fix the test_full_crawl errors you probably want to replace the mock_init function in test_full_crawl.py lines 96-102 with:
```python
        def mock_init(pipeline_cls, *_args, temp_dir=temp_dir, **_kwargs):
            pipeline_cls.api_url = None
            pipeline_cls.file_number = 1
            pipeline_cls.parent_file_path = temp_dir
            pipeline_cls.base_file_name = temp_dir / "output" / "all-links.csv"
            pipeline_cls.file_path = pipeline_cls.base_file_name
            pipeline_cls.current_file = open(pipeline_cls.file_path, "w", encoding="utf-8")
```			
or something similar.  I briefly looked at the test_url_file_size tests and noticed you are using the mocker fixture with pytests... did you mean to also add the pytest-mock plugin to the requirements file?  Besides that i didnt dig into it but seems like the mocks are not right.

File: cicd-scripts/app_install.sh
Comment (selfdanielj) [inline cicd-scripts/app_install.sh:1]: the other day when i went to run things I noticed we needed these three libs installed for our app to work.  Can you add them to the apt-get?
lzma
liblzma-dev
libbz2-dev

Comment (selfdanielj) [inline cicd-scripts/app_install.sh:1]: I think there was an issue with the `sudo pip install --upgrade` right? I was able to just remove the upgrade flag and have it work.

Comment (selfdanielj) [inline cicd-scripts/app_install.sh:1]: should there be a line to install playwright?  i don't think thats ever been in here.

File: cicd-scripts/app_start.sh
Comment (igoristic) [reply cicd-scripts/app_start.sh]: 🎣  Good catch! We are doing this check in the python code now.

File: cicd-scripts/app_install.sh
Comment (selfdanielj) [inline cicd-scripts/app_install.sh:97]: why do we need these? should they be added to requirement.txt if they are needed?

File: cicd-scripts/app_install.sh
Comment (selfdanielj) [inline cicd-scripts/app_install.sh:98]: i think you can probably just run `playwright install chromium` since that is what we are using in scrapy but this probably doesnt hurt.

File: cicd-scripts/app_install.sh
Comment (selfdanielj) [inline cicd-scripts/app_install.sh:85]: i thought the pip install usually happens inside the venv? do we need to upgrade the pip installed here?

File: cicd-scripts/helpers/run_without_ui.sh
Comment (selfdanielj) [inline cicd-scripts/helpers/run_without_ui.sh:3]: does this work? usually you would want to point to the python in the venv?

**Review by IsabelLaurenceau — APPROVED** [review]

File: search_gov_crawler/search_gov_spiders/pipelines.py
Comment (IsabelLaurenceau) [inline search_gov_crawler/search_gov_spiders/pipelines.py:68]: We need a way to identify the process /spider calling the file. This was done with my changes from https://github.com/GSA/searchgov-spider/pull/60 I recommend putting back my changes

**Review by IsabelLaurenceau — CHANGES_REQUESTED** [review]
See my comment on pipelines.py

File: search_gov_crawler/requirements.txt
Comment (igoristic) [reply search_gov_crawler/requirements.txt]: Good catch 👍

File: cicd-scripts/helpers/run_without_ui.sh
Comment (igoristic) [reply cicd-scripts/helpers/run_without_ui.sh]: Good call!

I fixed this by activating the virtual environment

File: cicd-scripts/app_install.sh
Comment (igoristic) [reply cicd-scripts/app_install.sh]: You're absolutely right

File: cicd-scripts/app_install.sh
Comment (igoristic) [reply cicd-scripts/app_install.sh]: I installed it with the `--with-deps` flag

File: tests/search_gov_spiders/test_deduplicator_pipeline.py
Comment (selfdanielj) [inline tests/search_gov_spiders/test_deduplicator_pipeline.py:1]: There was already some coverage of the de-deuplicator pipeline in `tests/search_gov_spiders/test_pipelines.py`... can you take a look and see if this can just replace that one?  I don't think we need tests in two places so maybe delete the other one if this fully replaces it?

Comment (igoristic) [reply tests/search_gov_spiders/test_deduplicator_pipeline.py:1]: Ah, you're right I completely missed the one in `test_pipeline.py`. Just, thought we weren't 100% covered on codec-climate

**Pull request merged**
